Updates (#5)
mbrandenburger/fabric-rfcs: text/0000-fabric-private-chaincode.md

Overall, FPC can be considered an addition to Fabric wherein all chaincode computation relies only on the correctness of data provided by an authenticated Client or computed inside of and signed by a validated Trusted Execution Environment. It assumes that the Ordering Service is a trusted system element. The Endorsing Peer outside of the Trusted Execution Environment, however, is considered untrusted.

> Writing chaincode for FPC should come natural to developers familiar with Fabric as the programming model (e.g., chaincode lifecycle, chaincode invocations and state) is the same as for normal Fabric chaincode.  The main differences are a (for now at least) different programming language (C++) and a shim API which, on the one hand, is somewhat simplified but, on the other hand, also contains FPC-specific extensions such as allowing for private as well as public ledger state as well as a secure (yet “replicable”) random-number generator. A link is provided below in the "FPC Chaincode Library / Shim" section to detailed documentation. 

Reviewer: Would be nice to resurrect the policy we talked about beforehand to aim to moving to a single-line per sentence while we change text.  Not that the current mode is really a big deal as using rich diff you can see the changes, just makes adding comments to the review a bit more cumbersome. 
PS: Actually, for this PR the rich diff doesn't work for the transaction flows, so it quite a bit more cumbersome to see the actual diff ....

> Writing chaincode for FPC should come natural to developers familiar with Fabric as the programming model (e.g., chaincode lifecycle, chaincode invocations and state) is the same as for normal Fabric chaincode.  The main differences are a (for now at least) different programming language (C++) and a shim API which, on the one hand, is somewhat simplified but, on the other hand, also contains FPC-specific extensions such as allowing for private as well as public ledger state as well as a secure (yet “replicable”) random-number generator. A link is provided below in the "FPC Chaincode Library / Shim" section to detailed documentation. 

Reviewer: Can we make this a "clickable" link to that section?

## Principal Components of FPC:

![Architecture](../images/FPC-Architecture.png) 

FPC Chaincode: A chaincode created by a developer to run in a Chaincode Enclave. Unlike regular Fabric chaincode, an FPC Chaincode must currently be written in C++ using our FPC SDK. A future goal for the project is to support additional languages, e.g., by the use of WebAssembly.

Chaincode Enclave: An enclave in which a single particular chaincode executes.  The Chaincode Enclave contains the chaincode (called an FPC Chaincode) to be executed, with its linked Chaincode Library. The FPC runtime is responsible for creating the Chaincode Enclave when the FPC chaincode is installed on the endorsing peers.

> Chaincode Library (also known as the FPC Shim): This is a shim interface exposed to an FPC Chaincode.  The Chaincode Library contains two parts, one residing inside the enclave and another one outside the enclave. The shim interface exposed to the FPC chaincode is written primarily in C++, whereas the counterpart is written in Go and mainly responsible for communication between the Peer and the FPC chaincode. The Chaincode Library enables an FPC Chaincode to talk to the Peer, in particular to invoke the chaincode and provide "normal" shim operations such as getState and putState. The FPC SDK provides the chaincode library to enable FPC chaincode developers to easily link to this library when compiling their FPC Chaincode and building a deployment artifact. It implements the interface from the Peer to the FPC Chaincode, encrypts and decrypts state and arguments, and implements Attestation. A link is provided below in the "FPC Chaincode Library / Shim" section to detailed documentation.

Reviewer: Same as above: would be nice to make this a clickable reference to the Chaincode library section

FPC Validator: This validation complements the Peer validation logic by validating transactions produced by an FPC Chaincode. It does this by checking that each transaction has been cryptographically signed (endorsed) by a Chaincode Enclave that is listed in the FPC Registry and is authorized for that chaincode. 

> FPC Client SDK extension: This extends the NodeSDK (or Go SDK) with extra functionality that allows users to write end-to-end secure FPC-based applications. 

Reviewer: I would keep it here generic; the MVP discussion below clarifies. (Otherwise we would also have to add here the cli given that we put it under this bucket.  There are also other client sdks, e.g., java, and the general discussion still would apply to them as well.)
```suggestion
FPC Client SDK extension: This extends the Fabric Client SDKs with extra functionality that allows users to write end-to-end secure FPC-based applications. 
```

> In particular, the client SDK extension will provide these core functions: 1) FPC transaction proposal creation (including transparent encryption of arguments) 2) FPC transaction proposal response validation and decryption of the result; 3) FPC deployment/setup functionality.

Reviewer: What is point 3) this sounds more like the admin commands, not client sdk? I.e., i would drop it ...

Author: removed

The encryption of transaction arguments in the proposal is performed by the Client SDK "under the covers" without requiring any special action by the users.

> With FPC the users still use normal invoke/query functions to issue transaction

Reviewer: ```suggestion
The encryption of transaction arguments in the proposal is performed by the Client SDK "under the covers" without requiring any special action by the users, i.e.,
the users still use normal invoke/query functions to issue FPC transaction
```

	When the particular chaincode function invoked finishes executing, the Chaincode Library produces a cryptographic signature over the input arguments, the read-write set, and the result. This is conceptually similar to the endorsement signature produced by the endorsing peer but instead of being rooted in an organizational entity like the peer it is based on the hardware and code entity and rooted in the attestation. Note that for this reason the Chaincode Library keeps track of the read/writeset during the invocation inside the Chaincode Enclave. Then the Chaincode library returns the result along with the signature back to the Peer. The Peer then uses the FPC signature as endorsement signature for the proposal response and sends it back to the Client.

* Step 3: Endorsement

> 	![Endorsement](../images/FPC-Endorsement.png)

Reviewer: Adding it here as i cannot add a comment directly to the .png (or at least don't know how to do it :-)

We should also add the source of this (and other .pngs) to version control so can easily change them if necessary in the future ...

Author: Agreed! @jrlinton can you please create PR to push the power point file to the FPC repo?

* Step 6: Trusted Ledger Revalidation

	![Revalidate](../images/FPC-Revalidate.png)

	In addition to the standard commitment process, the Peer also forwards all validated blocks to the Ledger Enclave in order to establish a full and current trusted view of the ledger. The same validation steps described above are repeated inside the Enclave, and then the transaction is committed to the trusted version of the Ledger inside this Enclave. This constitutes an update to the World State Integrity Metadata: for each Key Value Pair of World State, a second Key Value Pair is stored in the Trusted Ledger containing Integrity Metadata (a cryptographic hash value) along with channel-specific details needed to verify that the transaction was produced by a valid authorized participant in the channel.

* Step 7: Commitment

	![Commit](../images/FPC-Commit.png)

	If these validation processes succeed,the transaction is marked valid and is committed to the local ledger of the Peer. The Write Set is committed to the World State. This completes the transaction.

> ### Subsequent Query and Validation of World State Data by FPC Chaincodes

Reviewer: having here a ### section after lots of text seems a bit surprising, i guess it should be a ## section? (same for below encrypted elements?)

Related to sections, Terminology should also be a ## section. I guess we might have at some point just to do a systematic sectioning pass?

Author: ok, let's do an extra pass on sectioning later once we have "stable" content. That may be easier.

Subsequent to the transaction described above: When any FPC Chaincode accesses the World State, the Chaincode Library shim is responsible not only for passing the query to the Peer but also querying the Ledger Enclave for the Integrity Metadata for the Key in question. It performs a cryptographic hash of the Key Value Pair of World State returned by the Peer and checks it against the Integrity Metadata from the Ledger Enclave for that key. If the hashes do not match, it detects an integrity violation.

> ### Encrypted Elements

Reviewer: We might want to say here also who holds the private keys? (The key of course is that except for the results which are decryptable by the requestor, none of the other information is decryptable by anybody other than the chaincode enclaves (and the SGX HW for memory)

- FPC currently is installed on a Peer, and Endorsing Peers must use the FPC chaincode runtime. 
- FPC currently requires that the Endorsing Peers run on an Intel x86 machine with the SGX feature enabled. We plan to add support for other Trusted Execution Environments in future releases. 
- To simplify the development as much as possible and provide chaincode developers the same experience as with "traditional" chaincode development, FPC requires developers to use a chaincode SDK that abstracts the TEE specific implementation details as much as possible.
- FPC’s runtime relies on a external chaincode launcher, and its attestation infrastructure requires the installation of an FPC Registry chaincode in the channel
- FPC components running inside the enclave are implemented in C++; whereas components running outside are implemented in Go following Fabric programming standards/guidelines

## Fabric Features Not Yet Supported

In order to focus the development resources on the core components of FPC, the MVP excludes certain Fabric features, which will be added in the future.

> - Multiple Implementations for a Single Chaincode - this feature added in Fabric 2.0 is fundamentally incompatible with FPC's architecture: for chaincodes to be considered equivalent they must be bit-for-bit identical in order to generate matching identities (i.e. MRENCLAVE)

Reviewer: Trying to stress that this is not a FPC-specific limitation but more fundamental
```suggestion
- Multiple Implementations for a Single Chaincode -- a feature added in Fabric 2.0 -- is fundamentally incompatible for private chaincode like FPC where _all_ information flows have to be controlled: for FPC chaincodes to be considered equivalent they must be bit-for-bit identical in order to generate matching identities (i.e. MRENCLAVE)
```

Reviewer: You didn't like that suggestion? Or just overlooked it?

Author: overlooked, accepted

- Arbitrary endorsement policies
- Chaincode-to-chaincode invocations (cc2cc)
- State-based endorsement

> - Custom endorsement / validation plugins for FPC

Reviewer: reorder to make endorsement policy related stuff together
```suggestion
- Arbitrary endorsement policies
- State-based endorsement
- Custom endorsement / validation plugins for FPC
- Chaincode-to-chaincode invocations (cc2cc)
```
i guess rich db query (couchdb) also follows in this category?

In addition to the peer-side validation, clients must also be able to verify FPC transactions.
This process is similar to the validation step as described above; the FPC client SDK extension will implement this functionality and make this process thereby transparent to the users. A low level FPC client SDK API may also provide the functionality to the end user directly.

## Management API Changes

While the management API for Fabric is mostly unchanged, some modifications are needed for FPC to work. 
In particular, FPC extends the Fabric's lifecycle API with additional commands to create an FPC enclave and handle the key provisioning.

> These are detailed separately in this document: [Link](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/docs/design/fabric-v2%2B/fpc-management.md)

Reviewer: ```suggestion
These are detailed separately in the [FPC Management API document](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/docs/design/fabric-v2%2B/fpc-management.md)
```

Reviewer: You accepted the related suggestion for shim but not this one. I guess you overlooked it?

Author: thanks, accepted

## FPC Chaincode Library / Shim

As described above, FPC follows the programming model used in the standard Fabric Go shim and offers a C++ based Chaincode Library / Shim to FPC chaincode developers. 
For MVP, the FPC Shim comprises a subset of the standard Fabric Shim and is complemented in the future.

> These details are documented separately in the Shim header file itself: [Link](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/ecc_enclave/enclave/shim.h)

Reviewer: ```suggestion
These details are documented separately in the Shim header file itself: **[ecc_enclave/enclave/shim.h](https://github.com/hyperledger-labs/fabric-private-chaincode/blob/master/ecc_enclave/enclave/shim.h)**
```

## TEE platform support

Currently, our FPC Runtime and the SDK focuses on Intel SGX SDK. However, components such as the FPC Registry are already designed to support attestations by other TEE platforms as they mature and gain remote attestation capabilities. Also, other components such as the Go part of the FPC chaincode library don't have an Intel SGX depency and can easily be reused. We plan to explore other TEE platforms such as AMD SEV in the future.

> ## A Opportunity for Endorsement Policies

Reviewer: ```suggestion
## An Opportunity for richer Endorsement Policies
```

- Trusted Ledger Validation: In order to make a trusted “oracle” of information on the current world state, we’ve essentially had to replicate the Peer’s existing validation mechanism inside an Enclave, in C++ - an inefficient duplication of functionality that could lead to version synchronization problems. We hope to explore a path toward a modular validation mechanism that would eliminate this redundancy; possibly enabled by recent work on supporting Go inside the TEE.

- Attestation-Based Endorsements: In order to integrate cryptographic Attestation of the TEE with Fabric endorsements, we created custom endorsement and validation plugins and include the enclave signature as part of the response payload. This added payload and the need for every Peer to run a custom validation plugin seem undesirable in the long term. Our current plan is to create a custom Membership Service Provider to replace the original mechanism. The new TEE MSP type would treat TEEs as members of an Org, replacing the Peer’s endorsing signature with a TEE Signature. We believe this approach would be more elegant and flexible, and would lay the ground for easier enablement of new TEE types.

> - Go plugins for Trusted Ledger and Validation: In order to build FPC without recompiling the Peer, the current implementation makes use of Go plugins to enable our Trusted Ledger and Validation functions in the Peer. We understand that Go plugins may not be viable long-term; if they were no longer to be supported, we would have to recompile the Peer.

Reviewer: Well, we already _do_ recompile the peer for TLCC/trusted_ledger. I guess here the issue is more that here the patch in the peer might be more complicated than for tlcc (although we probably don't know as so far there was no incentive to really investigate?)

Author: updated

The FPC team welcomes the community’s advice on how each of these touch-points to Fabric should be handled going forward, and hope to solidify our plans for each element through the RFC process.

> ## Roadmap Features

Reviewer: This list might be read as a priority list and/or what is likely to come first after MVP.  But right now certainly not in the low-hanging fruit order :-)

Reviewer: Another feature which folks very likely will ask would seem CC2CC?

Author: added

- WebAssembly Chaincode: A primary goal for FPC moving forward is to support WebAssembly chaincode, and by extension all languages that compile to WASM.
There has already been extensive development of a high-performance open source WASM Interpreter / Compiler for Intel SGX Enclaves in the Private Data Objects project, and our current plan is to adopt that capability in the next major phase of FPC.
FPC's modular architecture has been designed from the beginning to enable this drop-in capability.

- Other TEEs: The FPC team is also participating in early discussions in the Confidential Computing Consortium, which aims to provide a standardized way of deploying WASM across multiple TEE technologies.
We hope to leverage this work when extending FPC to support these other TEEs.

- Endorsing Policies: The current version supports only a single designated FPC Endorsing Peer; it is our intention to support multiple FPC Endorsing Peers in the upcoming MVP release.
In future releases this would enable rich Endorsement Policies as described above.

> - Deployment Policies: We intend to support deployment policies which allow the users to define where a FPC chaincode is allowed to be deployed.  For instance, the user could define that a certain FPC chaincode can only executed by an endorsing peer on a on-premise node of a certain organization.  This allows enhanced risk management.

Reviewer: Maybe put Risk Management more promising, e..g., label as Risk Management & Deployment Policies?

Overall, FPC can be considered an addition to Fabric wherein all chaincode computation relies only on the correctness of data provided by an authenticated Client or computed inside of and signed by a validated Trusted Execution Environment. It assumes that the Ordering Service is a trusted system element. The Endorsing Peer outside of the Trusted Execution Environment, however, is considered untrusted.

> Writing chaincode for FPC should come natural to developers familiar with Fabric as the programming model (e.g., chaincode lifecycle, chaincode invocations and state) is the same as for normal Fabric chaincode.  The main differences are a (for now at least) different programming language (C++) and a shim API which, on the one hand, is somewhat simplified but, on the other hand, also contains FPC-specific extensions such as allowing for private as well as public ledger state as well as a secure (yet “replicable”) random-number generator. A link is provided below in the "FPC Chaincode Library / Shim" section to detailed documentation. 

Reviewer: I guess you also need a clickable suggestion for the clickable link ;-?
```suggestion
Writing chaincode for FPC should come natural to developers familiar with Fabric as the programming model (e.g., chaincode lifecycle, chaincode invocations and state) is the same as for normal Fabric chaincode.  The main differences are a (for now at least) different programming language (C++) and a shim API which, on the one hand, is somewhat simplified but, on the other hand, also contains FPC-specific extensions such as allowing for private as well as public ledger state as well as a secure (yet “replicable”) random-number generator. See more information in the [FPC Chaincode Library / Shim](#fpc-chaincode-library--shim) section. 
```

## Principal Components of FPC:

![Architecture](../images/FPC-Architecture.png) 

FPC Chaincode: A chaincode created by a developer to run in a Chaincode Enclave. Unlike regular Fabric chaincode, an FPC Chaincode must currently be written in C++ using our FPC SDK. A future goal for the project is to support additional languages, e.g., by the use of WebAssembly.

Chaincode Enclave: An enclave in which a single particular chaincode executes.  The Chaincode Enclave contains the chaincode (called an FPC Chaincode) to be executed, with its linked Chaincode Library. The FPC runtime is responsible for creating the Chaincode Enclave when the FPC chaincode is installed on the endorsing peers.

> Chaincode Library (also known as the FPC Shim): This is a shim interface exposed to an FPC Chaincode.  The Chaincode Library contains two parts, one residing inside the enclave and another one outside the enclave. The shim interface exposed to the FPC chaincode is written primarily in C++, whereas the counterpart is written in Go and mainly responsible for communication between the Peer and the FPC chaincode. The Chaincode Library enables an FPC Chaincode to talk to the Peer, in particular to invoke the chaincode and provide "normal" shim operations such as getState and putState. The FPC SDK provides the chaincode library to enable FPC chaincode developers to easily link to this library when compiling their FPC Chaincode and building a deployment artifact. It implements the interface from the Peer to the FPC Chaincode, encrypts and decrypts state and arguments, and implements Attestation. A link is provided below in the "FPC Chaincode Library / Shim" section to detailed documentation.

Reviewer: ```suggestion
Chaincode Library (also known as the FPC Shim): This is a shim interface exposed to an FPC Chaincode.  The Chaincode Library contains two parts, one residing inside the enclave and another one outside the enclave. The shim interface exposed to the FPC chaincode is written primarily in C++, whereas the counterpart is written in Go and mainly responsible for communication between the Peer and the FPC chaincode. The Chaincode Library enables an FPC Chaincode to talk to the Peer, in particular to invoke the chaincode and provide "normal" shim operations such as getState and putState. The FPC SDK provides the chaincode library to enable FPC chaincode developers to easily link to this library when compiling their FPC Chaincode and building a deployment artifact. It implements the interface from the Peer to the FPC Chaincode, encrypts and decrypts state and arguments, and implements Attestation.  See more information in the [FPC Chaincode Library / Shim](#fpc-chaincode-library--shim) section..
```

* Step 6: Trusted Ledger Revalidation

	![Revalidate](../images/FPC-Revalidate.png)

	In addition to the standard commitment process, the Peer also forwards all validated blocks to the Ledger Enclave in order to establish a full and current trusted view of the ledger. The same validation steps described above are repeated inside the Enclave, and then the transaction is committed to the trusted version of the Ledger inside this Enclave. This constitutes an update to the World State Integrity Metadata: for each Key Value Pair of World State, a second Key Value Pair is stored in the Trusted Ledger containing Integrity Metadata (a cryptographic hash value) along with channel-specific details needed to verify that the transaction was produced by a valid authorized participant in the channel.

* Step 7: Commitment

	![Commit](../images/FPC-Commit.png)

	If these validation processes succeed,the transaction is marked valid and is committed to the local ledger of the Peer. The Write Set is committed to the World State. This completes the transaction.

> ### Subsequent Query and Validation of World State Data by FPC Chaincodes

Reviewer: Let's at least fix this one?
```suggestion
## Subsequent Query and Validation of World State Data by FPC Chaincodes
```

Subsequent to the transaction described above: When any FPC Chaincode accesses the World State, the Chaincode Library shim is responsible not only for passing the query to the Peer but also querying the Ledger Enclave for the Integrity Metadata for the Key in question. It performs a cryptographic hash of the Key Value Pair of World State returned by the Peer and checks it against the Integrity Metadata from the Ledger Enclave for that key. If the hashes do not match, it detects an integrity violation.

> ### Encrypted Elements

Reviewer: ```suggestion
## Encrypted Elements
```

![Encryption](../images/FPC-Encryption.png)

> Encrypted elements of the FPC architecture (over and above those in Fabric, such as TLS tunnels from Client to Peer) include:
> - The Arguments in a proposal message
> - The results of execution in a proposal response message returned to the Client
> - All contents of memory in both the Chaincode Enclave(s) and the Ledger Enclave
> - All FPC transaction read/writesets are written to the Ledger (by default).

Reviewer: I guess you wanted a clickable suggestion?
```suggestion
Encrypted elements of the FPC architecture (over and above those in Fabric, such as TLS tunnels from Client to Peer) include:
- The Arguments in a proposal message
- The results of execution in a proposal response message returned to the Client
- All contents of memory in both the Chaincode Enclave(s) and the Ledger Enclave
- All FPC transaction read/writesets are written to the Ledger (by default).
Note that with the exception of the results, where also a legitimate requestor knows the secret keys, all secret/private keys are known only by to the enclaves or, for memory encryption, to the HW memory encryption engine (with the hardware also enforcing that only legitimate enclaves have access to the unencrypted memory).
```
